File: kattis/safehouses/safehouses.cpp
#include <iostream>
#include <vector>
using namespace std;

typedef pair<int,int> ii;

char g[200][200];

int main(){
	int n;
	cin >> n;

	vector<ii> spies;
	vector<ii> houses;
	for (int i = 0; i < n; ++i){
		for (int j =0 ; j < n; ++j){
			cin >> g[i][j];
			if (g[i][j] == 'S')	spies.push_back({i,j});
			if (g[i][j] == 'H')	houses.push_back({i,j});
		}
	}

	int ans = 0;
	for (auto spy : spies){
		int best = 9999;
		for (auto house : houses){
			best = min(best, abs(spy.first - house.first) + abs(spy.second-house.second));
		}
		ans = max(ans, best);
	}

	cout << ans << endl;

	return 0;
}

File: kattis/safehouses/safehouses.py

n = int(input())
grid = []
for i in range(n):
	grid.append(input().strip())


s = []
h = []
for i in range(n):
	for j in range(n):
		if grid[i][j] == 'S':
			s.append((i, j))
		elif grid[i][j] == 'H':
			h.append((i,j))


ans = 0
for spy in s:
	closest = 99999
	for house in h:
		dist = abs(spy[0] - house[0]) + abs(spy[1] - house[1])
		closest = min(closest, dist)
	ans = max(ans, closest)


print(ans)


